src/Expander/TC.hs

```diff
@@ -102,22 +100,29 @@ inst :: Scheme Ty -> [Ty] -> Expand Ty
 inst (Scheme n ty) ts
   | length ts /= fromIntegral n =
     throwError $ InternalError "Mismatch in number of type vars"
-  | otherwise = instNorm ty
+  | otherwise = instTy ty
   where
-    instNorm t = do
+    instTy :: Ty -> Expand Ty
+    instTy t = do
       t' <- normType t
-      Ty <$> inst' (unTy t')
+      Ty <$> instTyF (unTy t')
+
+    instTyF :: TyF Ty -> Expand (TyF Ty)
+    instTyF (TyF ctor tArgs) = do
+      let TyF ctor' tArgsPrefix = instCtor ctor
+      tArgsSuffix <- traverse instTy tArgs
+      pure $ TyF ctor' (tArgsPrefix ++ tArgsSuffix)
```

Comment:
The fact that we have to concatenate two lists is new.

Previously, the types were guaranteeing that a `TSchemaVar` is never applied to type arguments. So when we encountered a `TSchemaVar`, that was it, we returned `ts !! fromIntegral i` and we did not have any type arguments to recur on.

But now, according to the types, a `TSchemaVar` can have type arguments. So when `instCtor` encounters a `TSchemaVar`, it once again returns `ts !! fromIntegral` as before, and that might be e.g. a `TFun` with two type arguments. Or maybe `instCtor` encounters a `TFun`, and the two type arguments are instead in `tArgs` (`tArgsSuffix` after they are instantiated). So we have two sources of type arguments which we need to combine into one.

In practice, a `TSchemaVar` will not have type arguments, because higher-kinded types are not yet supported. So one of the two lists will be empty, and `++` is a complicated way to pick the non-empty one. Thinking a bit ahead, however, once we do support higher-kinded types, we might have `f Int` with `f` expanding to `Either String`; in this case, we want to concatenate `[String]` with `[Int]` to get `[String, Int]`, that is, `Either String Int`. So while it currently doesn't matter whether we write `tArgsPrefix ++ tArgsSuffix` or `tArgsSuffix ++ tArgsPrefix`, once we have higher-kinded types, we will want `tArgsSuffix ++ tArgsPrefix`, so we might as well use that now.

Reply:
This makes sense. But why is the comment here, rather than in the source file?

Reply:
because I think comments which explain why a _change_ was made belong either in commit messages or PR comments. I can turn it into a comment which explains why the code _does_ what it does if you like.

Reply:
added.

Reply:
Looks good!